### PR TITLE
podman generate spec

### PR DESCRIPTION
**Pull request by cdoern**
implement a new command `podman generate spec` which can formulate a json specgen to be consumed by both the pod
and container creation API.

supported flags are

--verbose (default true) print output to the terminal
--compact print the json output in a single line format to be piped to the API
--filename put the output in a file
--clone rename the pod/ctr in the spec so it won't conflict w/ an existing entity

Signed-off-by: Charlie Doern <cdoern@redhat.com>

#### Does this PR introduce a user-facing change?

```release-note
introduce podman generate spec which allows users to formulate Specgen data into a json file.
```

### CODE REVIEW

**Comment by cdoern** [comment]
@rhatdan @baude @vrothberg @mheon PTAL

**Comment by cdoern** [comment]
I think all of the failures are from the API tests hard coding how many containers should exist, I will add a line to remove the one I create and then fix the shell completion

File: docs/source/markdown/podman-generate-spec.1.md
Comment (rhatdan) [inline docs/source/markdown/podman-generate-spec.1.md:19]: Do you think --clone should have -c?

File: cmd/podman/generate/spec.go
Comment (rhatdan) [inline cmd/podman/generate/spec.go:41]: Print should be capitalized.

File: cmd/podman/generate/spec.go
Comment (rhatdan) [inline cmd/podman/generate/spec.go:44]: Print should be capitalized.

File: cmd/podman/generate/spec.go
Comment (rhatdan) [inline cmd/podman/generate/spec.go:47]: Should this just be --name?  Rather then podman generating a random name?

File: docs/source/markdown/podman-generate-spec.1.md
Comment (rhatdan) [inline docs/source/markdown/podman-generate-spec.1.md:27]: Why do we need this flag?  If I specify --filename it should not go to screen.

File: docs/source/markdown/podman-generate.1.md
Comment (rhatdan) [inline docs/source/markdown/podman-generate.1.md:17]: I thought we were going to generate YAML?

File: pkg/domain/infra/tunnel/generate.go
Comment (rhatdan) [inline pkg/domain/infra/tunnel/generate.go:49]: Reason for this?

File: pkg/specgen/generate/container.go
Comment (rhatdan) [inline pkg/specgen/generate/container.go:545]: If you force users to name their pods/containers, then none of this is necessary.

File: pkg/specgen/generate/pod_create.go
Comment (rhatdan) [inline pkg/specgen/generate/pod_create.go:294]: What does this code do, it looks like it marshals and unmarshals the same thing?

**Review by rhatdan — APPROVED** [review]

File: docs/source/markdown/podman-generate.1.md
Comment (cdoern) [reply docs/source/markdown/podman-generate.1.md:17]: the API takes JSON. I agree YAML is more ideal for readability but has no use case for how we aim for this to be applied.

Comment (rhatdan) [reply docs/source/markdown/podman-generate.1.md:17]: Is it easy to convert to YAML?  If so could we add a flag to convert it?

File: docs/source/markdown/podman-generate-spec.1.md
Comment (Luap99) [reply docs/source/markdown/podman-generate-spec.1.md:27]: I agree we should not need this flag

File: pkg/domain/infra/abi/generate.go
Comment (Luap99) [inline pkg/domain/infra/abi/generate.go:119]: Please do not use prints under pkg/..., output should be only handled in cmd/podman... for a consistent structure.
I would just return the byte array and then the cmd/podman can either write to a file or stdout.

File: docs/source/markdown/podman-generate-spec.1.md
Comment (rhatdan) [reply docs/source/markdown/podman-generate-spec.1.md:21]: Should we default to compact and force users to use | json_pp or equivalent?

Or if we generate YAML file by default, then --json would ouput in --compact format.

Comment (Luap99) [reply docs/source/markdown/podman-generate-spec.1.md:21]: @cdoern newlines are totally valid in json. You have an issue with your shell since you do not escape the output which is always a very bad idea. Try to put the `"$()"` in quotes.

File: pkg/specgen/generate/pod_create.go
Comment (cdoern) [reply pkg/specgen/generate/pod_create.go:294]: until I rework the Pod API, this needs to be done in order to make the podspecgen a viable option. since infra holds a lot of the information for pods, the pod spec is kind of empty with overlapping fields. This maps the overlapping fields back to the pod spec since we are not allowed to expose infra's spec in json at the moment.

File: docs/source/markdown/podman-generate-spec.1.md
Comment (cdoern) [reply docs/source/markdown/podman-generate-spec.1.md:27]: ok

File: cmd/podman/generate/spec.go
Comment (cdoern) [reply cmd/podman/generate/spec.go:47]: sure, that works for me

**Comment by cdoern** [comment]
/hold

waiting until @baude and the rest of the team decide specific on this in the coming weeks.

File: cmd/podman/generate/spec.go
Comment (vrothberg) [inline cmd/podman/generate/spec.go:53]: I don't like the name requirement as the user faces the challenge of finding a unique name.  I think that's Podman's job.

Without having looked at the details, my intuitive expectation is that Podman would be clever enough to assign a name if none is set.  That would make such spec-files _portable_.

Similar to `generate systemd --new`, I imagine a "generate once deploy as often as you want" kind of workflow.  Enforcing a name choise would complicate the workflow quite a bit.

Somehow related: mandatory options/flags should be arguments.

**Review by vrothberg — COMMENTED** [review]
Nice work, @cdoern!

File: cmd/podman/generate/spec.go
Comment (rhatdan) [reply cmd/podman/generate/spec.go:53]: What does `podman generate kube` do? We should follow that precedent.

Comment (vrothberg) [reply cmd/podman/generate/spec.go:53]: `generate kube` does not alter the name(s) of pods/containers.  `kube play` will error out though when trying to create a pod/container with a conflicting name.

Comment (rhatdan) [reply cmd/podman/generate/spec.go:53]: So shouldn't this work the same, and force users to change the name?  IE If we are going to allow users to modify and ship the spec?

Comment (vrothberg) [reply cmd/podman/generate/spec.go:53]: I think that sounds like a sane default.

I also see beauty in not setting the name at all in the spec.  This way, the files are really portable as podman can create a unique name without the risk of a name conflict when the file is deployed on another machine.  I do not have a strong opinion whether clearing the name should be the default or not.  Probably worth adding a flag?

* `podman generate spec --name=false` -> clears the name
* `podman generate spec --name=true` -> generates with the name

But I am focusing on containers only.  I am not sure whether clearing names would work for pods.

In case the feature gets promoted for higher tasks we should check how we want it to work in automated fashion.

File: pkg/domain/entities/generate.go
Comment (cdoern) [reply pkg/domain/entities/generate.go:58]: good idea! I have not implemented the API endpoint in this PR, since i felt it was out of scope. However, future proofing always makes sense.

File: cmd/podman/generate/spec.go
Comment (cdoern) [reply cmd/podman/generate/spec.go:53]: I have a function used by clone that can generate a new name for both pods and containers. I will add the `--name=t/f` flag I think that makes the most sense.

it is basically the same as my `--clone` flag from a previous commit just with a different, more general name.

Comment (vrothberg) [reply cmd/podman/generate/spec.go]: What's important IMO is that we can generate a file once and deploy it anywhere without risking a name conflict on a node.  As long as Podman can auto-assign a name when _reading_/_running_ the spec, I am happy :)

Comment (vrothberg) [reply cmd/podman/generate/spec.go]: The crux of the matter is to assign names when applying/running/playing the file and not when generating it.

Comment (cdoern) [reply cmd/podman/generate/spec.go]: ok so @vrothberg @rhatdan, here an auto-generation might be warranted, but when if/when we implement a `run --spec` we can offer a flag to rename?

File: docs/source/markdown/podman-generate-spec.1.md
Comment (TomSweeneyRedHat) [inline docs/source/markdown/podman-generate-spec.1.md:12]:
```suggestion
This JSON can then be used as input for the podman API, specifically for Podman container and pod creation. Specgen is Podman's internal structure for formulating new container-related entities.
```

File: docs/source/markdown/podman-generate-spec.1.md
Comment (TomSweeneyRedHat) [inline docs/source/markdown/podman-generate-spec.1.md:18]:
```suggestion
Print the output in a compact, one line format. This is useful when piping the data to the Podman API
```

File: docs/source/markdown/podman-generate-spec.1.md
Comment (TomSweeneyRedHat) [inline docs/source/markdown/podman-generate-spec.1.md:26]: Pronouns are evil!   At least in man pages.  ;^)
```suggestion
Rename the pod or container, so it does not conflict with the existing entity. This is helpful when the JSON is to be used before the pod or container that is generating it is deleted.
```

File: cmd/podman/generate/spec.go
Comment (cdoern) [reply cmd/podman/generate/spec.go:17]: isn't this saying `container or pod`, that is what I assumed this meant

Comment (Luap99) [reply cmd/podman/generate/spec.go:17]: ... means zero or more arguments, remove the dots and this is fine

File: pkg/domain/infra/tunnel/generate.go
Comment (cdoern) [reply pkg/domain/infra/tunnel/generate.go:49]: I mean, I could add a remote endpoint but does this have a usecase remotely at the moment since the only usage is really directly piping to the API?

File: docs/source/markdown/podman-generate.1.md
Comment (cdoern) [reply docs/source/markdown/podman-generate.1.md:17]: @baude, thoughts on this? I've been letting this sit while you were away. personally I think just JSON makes the most sense for usage, but if we want to eventually support consumption by a cli command, yaml could work too.

File: cmd/podman/generate/spec.go
Comment (Luap99) [inline cmd/podman/generate/spec.go:44]: I still think this flag is unessesary. This is absolutely not necessary, just quote your arguments properly.

File: docs/source/markdown/podman-generate-spec.1.md
Comment (Luap99) [inline docs/source/markdown/podman-generate-spec.1.md:7]: Need to remove the ...

File: cmd/podman/generate/spec.go
Comment (cdoern) [reply cmd/podman/generate/spec.go:44]: odds are if I ran into this issue with my shell, others will too and we will get bug reports something that isn't broken on our end. Harmless safety IMO

**Comment by cdoern** [comment]
@containers/podman-maintainers need some reviews on this, it has been dangling for a while

**Comment by cdoern** [comment]
@containers/podman-maintainers PTAL

File: docs/source/markdown/podman-generate-spec.1.md
Comment (mheon) [inline docs/source/markdown/podman-generate-spec.1.md:26]: Seem to be missing the last sections of the file - examples, cross-references?

File: docs/source/markdown/podman-generate.1.md
Comment (mheon) [reply docs/source/markdown/podman-generate.1.md:17]: Any chance we can have this link to the API docs for Specgen?

File: cmd/podman/generate/spec.go
Comment (mheon) [inline cmd/podman/generate/spec.go:31]: Given this accepts both containers and pods, would flags to force one or the other (`podman generate spec --ctr` or similar) make sense?

File: pkg/domain/infra/tunnel/generate.go
Comment (mheon) [reply pkg/domain/infra/tunnel/generate.go:49]: I think it actually makes a lot of sense that an API user would want to get the API definition of an existing container?

File: pkg/specgen/generate/container.go
Comment (mheon) [reply pkg/specgen/generate/container.go:560]: Can you rename `kind` to `isCtr` or something similar?

File: cmd/podman/generate/spec.go
Comment (vrothberg) [reply cmd/podman/generate/spec.go:31]: `podman inspect` has a `--type` flag that I'd prefer for consistency.

But we could add that in a follow-up PR.

File: pkg/specgen/generate/container.go
Comment (vrothberg) [reply pkg/specgen/generate/container.go:560]: Still open.  It would be helpful to make the comment of the function more verbose such that new eyes won't have to read the function.

**Review by vrothberg — COMMENTED** [review]
Other than the mini nit, LGTM

Really nice work, @cdoern

**Review by giuseppe — APPROVED** [review]
LGTM

**Pull request merged**
